docs/source/format/Columnar.rst

```diff
@@ -0,0 +1,1266 @@
+      |-------------------------|-----------------------|
+      | 00001011                | 0 (padding)           |
+
+    * Children arrays:
+      * field-0 array (`List<char>`):
+        * Length: 4, Null count: 2
+        * Null bitmap buffer:
+
+          | Byte 0 (validity bitmap) | Bytes 1-63            |
+          |--------------------------|-----------------------|
+          | 00001001                 | 0 (padding)           |
+
+        * Offsets buffer:
+
+          | Bytes 0-19     |
+          |----------------|
+          | 0, 3, 3, 3, 7  |
+
+         * Values array:
+            * Length: 7, Null count: 0
+            * Null bitmap buffer: Not required
+
+            * Value buffer:
+
+              | Bytes 0-6      |
+              |----------------|
+              | joemark        |
+
+      * field-1 array (int32 array):
+        * Length: 4, Null count: 1
+        * Null bitmap buffer:
+
+          | Byte 0 (validity bitmap) | Bytes 1-63            |
+          |--------------------------|-----------------------|
+          | 00001011                 | 0 (padding)           |
+
+        * Value Buffer:
+
+          |Bytes 0-3   | Bytes 4-7   | Bytes 8-11  | Bytes 12-15 | Bytes 16-63 |
+          |------------|-------------|-------------|-------------|-------------|
+          | 1          | 2           | unspecified | 4           | unspecified |
+
+While a struct does not have physical storage for each of its semantic
+slots (i.e. each scalar C-like struct), an entire struct slot can be
+set to null via the null bitmap. Any of the child field arrays can
+have null values according to their respective independent null
+bitmaps. This implies that for a particular struct slot the null
+bitmap for the struct array might indicate a null slot when one or
+more of its child arrays has a non-null value in their corresponding
+slot.  When reading the struct array the parent null bitmap is
+authoritative.  This is illustrated in the example above, the child
+arrays have valid entries for the null struct but are 'hidden' from
+the consumer by the parent array's null bitmap.  However, when treated
+independently corresponding values of the children array will be
+non-null.
+
+Dense Union
+-----------
+
+A dense union is semantically similar to a struct, and contains an
+ordered sequence of types. While a struct contains multiple arrays, a
+union is semantically a single array in which each slot can have a
+different type.
+
+The union types may be named, but like structs this will be a matter
+of the metadata and will not affect the physical memory layout.
+
+We define two distinct union types that are optimized for different use
+cases. This first, the dense union, represents a mixed-type array with 5 bytes
+of overhead for each value. Its physical layout is as follows:
+
+* One child array for each type
+* Types buffer: A buffer of 8-bit signed integers, enumerated from 0 corresponding
```

Comment:
Are they really enumerated from 0? I thought those were logical type numbers, to allow for adding/removing types without rewriting the types buffer...